src/Illuminate/Database/Eloquent/Model.php

```diff
@@ -335,18 +336,23 @@ public static function clearBootedModels()
     /**
      * Register a new global scope on the model.
      *
-     * @param  \Illuminate\Database\Eloquent\ScopeInterface  $scope
+     * @param  \Illuminate\Database\Eloquent\ScopeInterface|\Closure  $scope
+     * @param  string|null  $identifier
      * @return void
      */
-    public static function addGlobalScope(ScopeInterface $scope)
+    public static function addGlobalScope($scope, $identifier = null)
     {
-        static::$globalScopes[get_called_class()][get_class($scope)] = $scope;
+        if (! $identifier) {
+            $identifier = $scope instanceof ScopeInterface ? get_class($scope) : uniqid('scope');
+        }
+
+        static::$globalScopes[get_called_class()][$identifier] = $scope;
```

Comment:
Ah, so this is where registration actually happens, and my first comment becomes incorrect.